include/up-transport-zenoh-cpp/ZenohUTransport.h

```diff
@@ -0,0 +1,102 @@
+
+#ifndef UP_TRANSPORT_ZENOH_CPP_ZENOHUTRANSPORT_H
+#define UP_TRANSPORT_ZENOH_CPP_ZENOHUTRANSPORT_H
+
+#include <up-cpp/transport/UTransport.h>
+
+#include <filesystem>
+#include <optional>
+
+namespace uprotocol::transport {
+
+/// @brief Zenoh implementation of UTransport
+///
+/// This implementation must meet the following requirements:
+///
+/// * [MUST] An implementation MUST support multiple simultaneous instantiatons
+///          from within the same application.
+///
+/// * [MAY]  An implimentation MAY require that each new instance within an
+///          application have a unique configuration from existing instances.
+///
+/// * [MUST] An implementation MUST allow for multiple users of the same
+///          instance.
+///
+/// * [MUST] An implementation MUST be thread-safe.
+///
+/// * [MUST] Throw an exception if the transport fails to initialize or the
+///          configuration is invalid.
+struct ZenohUTransport : public UTransport {
+	/// @brief Constructor
+	///
+	/// @param defaultUri Default Authority and Entity (as a UUri) for
+	///                   clients using this transport instance.
+	/// @param configFile Path to a configuration file containing the Zenoh
+	///                   transport configuration.
+	ZenohUTransport(const v1::UUri& defaultUri,
+	                const std::filesystem::path& configFile);
+
+	virtual ~ZenohUTransport() = default;
+
+protected:
+	/// @brief Send a message.
+	///
+	/// @param message UMessage to be sent.
+	///
+	/// @returns * OKSTATUS if the payload has been successfully
+	///            sent (ACK'ed)
+	///          * FAILSTATUS with the appropriate failure otherwise.
+	[[nodiscard]] virtual v1::UStatus sendImpl(
+	    const v1::UMessage& message) override;
+
+	/// @brief Represents the callable end of a callback connection.
+	using CallableConn = typename UTransport::CallableConn;
+
+	/// @brief Register listener to be called when UMessage is received
+	///        for the given URI.
+	///
+	/// @remarks If this doesn't return OKSTATUS, the public wrapping
+	///          version of registerListener() will reset the connection
+	///          handle before returning it to the caller.
+	///
+	/// @param sink_filter UUri for where messages are expected to arrive via
+	///                    the underlying transport technology. The callback
+	///                    will be called when a message with a matching sink
+	/// @param listener shared_ptr to a connected callback object, to be
+	///                 called when a message is received.
+	/// @param source_filter (Optional) UUri for where messages are expected to
+	///                      have been sent from. The callback will only be
+	///                      called for messages where the source matches.
+	///
+	/// @returns * OKSTATUS if the listener was registered successfully.
+	///          * FAILSTATUS with the appropriate failure otherwise.
+	[[nodiscard]] virtual v1::UStatus registerListenerImpl(
```

Comment:
@sashacmc, @evshary -- curious to hear your folks' thoughts on how to handle `registerListenerImpl` and `unregisterListenerImpl`.

From my reviews of `up-client-zenoh-rust`, Zenoh Query was used for doing uProtocol Request half of RPC upon send and it would be expected to have registered a listener for the Response message.

Would there be a check here and within the `sendImpl` function in order to properly dispatch "under the hood" to using the correct Zenoh communication mechanism?

How about the uP-L2 RpcClient? Currently within `up-client-zenoh-rust` this is handled by _not_ building on top of its `UTransport`, but instead being its own mechanism. I suppose we would move away from that concept, since the way uP-L1 and uP-L2 are within up-cpp, we would effectively "plug-in" the Zenoh implementation of UTransport into the up-cpp implementation of, say, RpcClient.

Sorry for the barrage of questions. Comes to mind as I read this. :slightly_smiling_face:

Reply:
> Would there be a check here and within the sendImpl function in order to properly dispatch "under the hood" to using the correct Zenoh communication mechanism?

I think so.

> How about the uP-L2 RpcClient? Currently within up-client-zenoh-rust this is handled by not building on top of its UTransport, but instead being its own mechanism. I suppose we would move away from that concept, since the way uP-L1 and uP-L2 are within up-cpp, we would effectively "plug-in" the Zenoh implementation of UTransport into the up-cpp implementation of, say, RpcClient.

The way uP-L2 using now is more intuitive, while uP-L1 needs to register response callback first and send the request. However, they should make no difference. I haven't taken a look at up-cpp, but I don't think it'll affect the interoperability.